### ISSUE TITLE
Ralph on FreeBSD-12 (current), TypeError ralph/data_importer/resources.py

### ISSUE THREAD

**Issue by daviddpd** [allegro/ralph#3134]
So, I'm working on getting Ralph running natively on FreeBSD ... which I did back in the end of May as a POC/demo, however I didn't take any notes, and it was hacked on over several days.   I've settled on Ralph as our "OpsDB" (DCIM), and wanted to get a cleaner install going.   (We are a hard core FreeBSD shop, and this is python ... so should be able to run anywhere.) 

I believe I've gotten past all the major FreeBSD side stuff, but I'm getting an python exception on an OO class TypeError.  Most likely I'm still off a version of some sub package, but both OO & python are my weak areas, so need at least a pointer or hint at where to look for this one.   Below is my backtrace as well as my diff and FreeBSD install directions. 

Once this is sorted out, I will make the native FreeBSD port/package and submit it, and likely maintain the port. 

```Python traceback
Traceback (most recent call last):
  File "/usr/local/bin/ralph", line 11, in <module>
    load_entry_point('ralph==3.0.0', 'console_scripts', 'ralph')()
  File "/usr/local/lib/python3.6/site-packages/ralph-3.0.0-py3.6.egg/ralph/__main__.py", line 29, in prod
    main('ralph.settings.prod')
  File "/usr/local/lib/python3.6/site-packages/ralph-3.0.0-py3.6.egg/ralph/__main__.py", line 14, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 328, in execute
    django.setup()
  File "/usr/local/lib/python3.6/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.6/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  File "/usr/local/lib/python3.6/site-packages/django/contrib/admin/apps.py", line 22, in ready
    self.module.autodiscover()
  File "/usr/local/lib/python3.6/site-packages/django/contrib/admin/__init__.py", line 24, in autodiscover
    autodiscover_modules('admin', register_to=site)
  File "/usr/local/lib/python3.6/site-packages/django/utils/module_loading.py", line 74, in autodiscover_modules
    import_module('%s.%s' % (app_config.name, module_to_search))
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 978, in _gcd_import
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load
  File "<frozen importlib._bootstrap>", line 950, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 205, in _call_with_frames_removed
  File "/usr/local/lib/python3.6/site-packages/ralph-3.0.0-py3.6.egg/ralph/assets/admin.py", line 35, in <module>
    from ralph.data_importer import resources
  File "/usr/local/lib/python3.6/site-packages/ralph-3.0.0-py3.6.egg/ralph/data_importer/resources.py", line 581, in <module>
    class BaseObjectsSupportRichResource(RalphModelResource):
TypeError: __class__ set to <class 'ralph.data_importer.resources.BaseObjectsSupportRichResource'> defining 'BaseObjectsSupportRichResource' as <class 'ralph.data_importer.resources.BaseObjectsSupportRichResource'>
```

# FreeBSD Patches 

Don't forget to apply this patch.  Super minor tweaks, but needed.

```diff
diff --git a/Makefile b/Makefile
index 9817af1d..653b97a6 100644
--- a/Makefile
+++ b/Makefile
@@ -13,7 +13,7 @@ upload-package:

 install-js:
        npm install
-       bower install
+       bower install --allow-root

 js-hint:
        find src/ralph|grep "\.js$$"|grep -v vendor|xargs ./node_modules/.bin/jshint;
diff --git a/requirements/base.txt b/requirements/base.txt
index 5d1acda2..e37ae317 100644
--- a/requirements/base.txt
+++ b/requirements/base.txt
@@ -26,5 +26,5 @@ rq==0.5.6
 six>=1.9.0
 sqlparse==0.1.16
 Unidecode==0.04.18
-tablib==0.11.2
+tablib==0.11.5
 factory-boy
```

# Package install and somewhat need order of operations. 

Using Python 3.6, which might be part of the errors, but only 2.7 and 3.6 have breath and depth in the FreeBSD ports collection. 

```shell 
# FreeBSD 12.0-CURRENT #1 r321002
# packages are from Jul 16, 2017, all default options buy of internal poudriere build. 
#
# SUDO /root installed :  I know bower doesn't want to run as root, however trying to install as a
#  user, ended up being much more complex, and was never able to get it this close.
# 
#  git clone ...
  cd ralph
  sudo pkg install gcc libsass py27-libsass npm git python36 py36-pip msgpack
  sudo ln -s  /usr/local/bin/pip-3.6 /usr/local/bin/pip3
  rehash # for you csh/tcsh people like me.
  sudo npm install bower 
  rehash 

# node-sass newest packages for FreeBSD are 3.9.  So it has to compile, 
# and clang/LLVM errors out on first file. 

  sudo env CXX=g++ env PATH=${PATH}:${HOME}/ralph/node_modules/bower/bin make install
  sudo python3.6 setup.py install
  rehash 
```
# helper run script  
 

```shell run.sh
#!/bin/sh

export DATABASE_NAME=ralph
export DATABASE_USER=ralph
export DATABASE_PASSWORD=XXXXX
export DATABASE_HOST=localhost
export RALPH_DEBUG=1

$@
```
# Run / bootstrap Ralph

```shell 
./run.sh ralph migrate
```

**Comment by vi4m** [comment]
@ar4s @mkurek have we tested ralph with python 3.6 already?

**Comment by daviddpd** [comment]
@vi4m  / @ar4s  / @mkurek  - indeed, it looks like it is a 3.6 vs 3.4 ... wow.  Rarely see things this picky about minor versions.   I believe I'm running the other non-documented setup with 3.5.

Since there are no py34- ports/packages ... the installed directions changed a tiny bit : 

```shell
  sudo pkg install -y curl gcc libsass py27-libsass npm git python34 msgpack
  curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
  sudo python3.4 get-pip.py
  sudo npm install bower 
  sudo env CXX=g++ env PATH=${PATH}:${HOME}/ralph/node_modules/bower/bin make install
  sudo python3.4 setup.py install
```
ralph migrate is still running at the moment ... but this looks promising.

(*) I do have a base set of other packages on these machines ... I'll try a completely bare FreeBSD install shortly. 

So - it works ... however, there are bunch of files missing ... mostly all the static stuff.   Previously, I couldn't find these files, so I cracked open one of the deb/egg's of ralph, and copied out the files.  Where is the step that is missing to get these files into the correct location ?  Running in debug mode, allows ralph to run ... but without javascript and css.

```shell
[ pts/0 app3.ixystems.com:~ ]
[ dpd ]> sh ./run.sh ralph createsuperuser
Username: admin
Email address: admin@ixsystems.com
Password:
Password (again):
Superuser created successfully.

[ pts/0 app3.ixystems.com:~ ]
[ dpd ]> uname -a
FreeBSD app3.ixystems.com 12.0-CURRENT FreeBSD 12.0-CURRENT #1 r321002: Sat Jul 15 01:45:38 UTC 2017     dpd@ixsystems.com:/usr/obj/usr/src/sys/IX  amd64

[ pts/0 app3.ixystems.com:~ ]
[ dpd ]> sh ./run.sh ralph runserver 0.0.0.0:8000  ## RALPH_DEBUG=1
[... bunch of django warnings ...] 
Performing system checks...

System check identified no issues (0 silenced).
August 06, 2017 - 14:21:40
Django version 1.8.18, using settings 'ralph.settings.prod'
Starting development server at http://0.0.0.0:8000/
Quit the server with CONTROL-C.
[06/Aug/2017 14:21:49] "GET /login/?next=/ HTTP/1.1" 200 8104
[06/Aug/2017 14:21:49] "GET / HTTP/1.1" 302 0
[06/Aug/2017 14:21:49] "GET / HTTP/1.1" 302 0
[06/Aug/2017 14:21:49] "GET /login/?next=/ HTTP/1.1" 200 8104
[06/Aug/2017 14:21:49] "GET /login/?next=/ HTTP/1.1" 200 8104
[06/Aug/2017 14:21:49] "GET /static/admin/js/jquery.js HTTP/1.1" 304 0
[06/Aug/2017 14:21:49] "GET /static/admin/js/core.js HTTP/1.1" 304 0
[06/Aug/2017 14:21:49] "GET /static/admin/js/jquery.init.js HTTP/1.1" 304 0
[06/Aug/2017 14:21:49] "GET /static/vendor/css/foundation-datepicker.css HTTP/1.1" 404 1730
[06/Aug/2017 14:21:49] "GET /static/vendor/js/webcomponents-lite.js HTTP/1.1" 404 1715
[06/Aug/2017 14:21:49] "GET /static/vendor/css/foundation-datepicker.css HTTP/1.1" 404 1730
[...]
```

**Comment by vi4m** [comment]
If you don't want to run in debug mode, the comand which copies all required static files is : `ralph collectstatic`.  It should be then served with nginx/apache. It could be troublesome for some of your not so skilled users, because it requires some (really minimal) nginx/apache setup to serve these static files, though. That's why I'm thinking about 3'rd alternative http://whitenoise.evans.io/en/stable/ What do you guys think ? @mkurek @ar4s ? Whitenoise allows to serve static files directly from within an app, it doesn't require debug mode on, and it works without external nginx/apache setup.

**Comment by daviddpd** [comment]
I haven't gotten down to used Ralph much yet, but seems to be functional with my last install directions on FreeBSD,  however, this is still ripped out static files from the DEB package.  I downloaded a package from:
 - https://dl.bintray.com/vi4m/ralph/dists/wheezy/main/binary-amd64/

Then tar'ed up `./opt/ralph/ralph-core/lib/python3.4/site-packages/ralph/admin/static/` and `./opt/ralph/ralph-core/lib/python3.4/site-packages/ralph/static` and untar'ed them into the FreeBSD location of `/usr/local/lib/python3.4/site-packages/ralph-3.0.0-py3.4.egg/ralph/admin/static` and `/usr/local/lib/python3.4/site-packages/ralph-3.0.0-py3.4.egg/ralph/static` - the built in default Django server hands out the statics from locations just fine.  I don't think there is a need for yet-another web server.

The `ralph collectstatic` ended up placing the static files in `/usr/local/lib/python3.4/site-packages/ralph-3.0.0-py3.4.egg/ralph/var/static` but I still have the other two static locations populated.     It looks like Django is still serving out those the admin/static and static/ locations.  I'm trying to do an apache reverse Proxy, which is working, but still working through the configs.

A little more documentation on some settings would be additionally helpful, these are likely Django specific but would be helpful for those not familiar with python and this framework.

One of them, would be how to configure base URLs.  I originally was a attempting with Reverse Proxy to place Ralph at `https://SERVERNAME/ralph` ... where Apache was handling the SSL, and passing HTTP over localhost.  However, Django/Ralph got confused at the base paths.  I tried a few things, but couldn't find the right settings.  I went ahead and did a name-based virtual host.  I'm also trying to use REMOTE_USER auth pass thru, but will open another ticket on that (I know I like single subject tickets.)

As far as this issue goes the Ralph DCIM & Asset Management System on FreeBSD ... works.  I'll be making a package as soon as I finish up auth.  I don't care who servers the static content, as long as I can get it installed into the correct places without hacks.

So, I'm gonna close this issue, as the root cause was indeed Python 3.6, and using python 3.4 solved it.
